### PR TITLE
Refine auto spill for fine grained executor

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: ref #7738

Problem Summary:
When fine grained shuffle is enabled, due to implicit depentency of all the operator belongs to the same TiDB executor, auto spill will meet a big performance regression if auto spill treated all the operator spill contexts from the same TiDB executor independently. This pr add `FineGrainedOperatorSpillContext` and try to align the spill behavior for fine grained executor.
### What is changed and how it works?
- Add `FineGrainedOperatorSpillContext`
- For operator that comes from the same TiDB executor, auto spill will be triggered at TiDB executor level, instead of triggering each operator independently

Test on TPCH-100 dataset, with 2 TiFlash node
For fine grained agg, test sql is
```
select /*+ mpp_1phase_agg() */ l_orderkey, max(L_COMMENT), max(L_SHIPMODE), max(L_SHIPINSTRUCT), max(L_SHIPDATE), max(L_EXTENDEDPRICE) from lineitem group by l_orderkey having sum(l_quantity) > 314
``` 
per node memory usage limit to 12G:

| without this pr | with this pr| 
|-------------|--------------|
| ~45s            | ~33s |

per node memory usage limit to 7G:

| without this pr | with this pr| 
|-------------|--------------|
| ~37s            | ~30s |

For fine grained sort, test sql is
```
select * from (select L_ORDERKEY,L_COMMENT,L_SHIPINSTRUCT, row_number() over(partition by l_orderkey order by l_orderkey) as rn from lineitem) d where l_orderkey + rn < 50;
```
per node memory usage limit to 12G:

| without this pr | with this pr| 
|-------------|--------------|
| ~45s            | ~24s |

per node memory usage limit to 7G:

| without this pr | with this pr| 
|-------------|--------------|
| ~45s            | ~25s |
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Core/FineGrainedOperatorSpillContext.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+/// with fine grained shuffle enable. When fine grained shuffle is enabled, executors like sort and aggregation
+/// will have n independent operators in TiFlash side, each of them have their own operator spill context. However
+/// these n operators are not really independent: they share the same source operator(exchange receiver), if any one
+/// of the operator stop consuming data from exchange receiver(for example, begin spill data), all the others will
+/// be stuck because the exchange receiver has a bounded queue for all the operators, any one of the operator stop
+/// consuming will make the queue full, and no other data can be pushed to exchange receiver. If all the n operator
+/// is triggered to spill serially, it will affects the overall performance seriously.
+/// FineGrainedOperatorSpillContext is used to make sure that all the operators belongs to the same executor
+/// will be triggered to spill almost at the same time
+class FineGrainedOperatorSpillContext : public OperatorSpillContext
```
Comment (yibin87) [inline dbms/src/Core/FineGrainedOperatorSpillContext.h:31]: If we don't intend to inherit this class, better add final keyword.

File: dbms/src/Core/FineGrainedOperatorSpillContext.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+{
+protected:
+    std::vector<OperatorSpillContextPtr> operator_spill_contexts;
```
Comment (yibin87) [inline dbms/src/Core/FineGrainedOperatorSpillContext.h:34]: why not private field member?

File: dbms/src/Core/FineGrainedOperatorSpillContext.h (hunk 1 above)
Comment (windtalker) [reply dbms/src/Core/FineGrainedOperatorSpillContext.h]: Done

File: dbms/src/Core/FineGrainedOperatorSpillContext.h (hunk 2 above)
Comment (windtalker) [reply dbms/src/Core/FineGrainedOperatorSpillContext.h:34]: Done

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/Core/FineGrainedOperatorSpillContext.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    return ret;
+}
+
+Int64 FineGrainedOperatorSpillContext::triggerSpillImpl(Int64 expected_released_memories)
+{
+    Int64 original_expected_released_memories = expected_released_memories;
+    for (auto & operator_spill_context : operator_spill_contexts)
+    {
+        if (expected_released_memories <= 0)
+            operator_spill_context->triggerSpill(original_expected_released_memories);
```
Comment (SeaRise) [inline dbms/src/Core/FineGrainedOperatorSpillContext.cpp:53]: Why is this needed? Does this mean that more data than `expected_released_memories` will be spilled.

File: dbms/src/Flash/Planner/Plans/PhysicalTopN.cpp [hunk 4]
```diff
@@ -79,7 +79,7 @@ void PhysicalTopN::buildPipelineExecGroupImpl(
     // TODO find a suitable threshold is necessary; 10000 is just a value picked without much consideration.
     if (group_builder.concurrency() * limit <= 10000)
     {
-        executeLocalSort(exec_context, group_builder, order_descr, limit, context, log);
+        executeLocalSort(exec_context, group_builder, order_descr, limit, false, context, log);
```
Comment (SeaRise) [inline dbms/src/Flash/Planner/Plans/PhysicalTopN.cpp:82]: It looks like this is similar to fine grained.
How about using the same spilling strategy?

File: dbms/src/Core/FineGrainedOperatorSpillContext.cpp (hunk 3 above)
Comment (windtalker) [reply dbms/src/Core/FineGrainedOperatorSpillContext.cpp:53]: Yes, because the worst case is each operator is triggered to spill serially.

File: dbms/src/Flash/Planner/Plans/PhysicalTopN.cpp (hunk 4 above)
Comment (windtalker) [reply dbms/src/Flash/Planner/Plans/PhysicalTopN.cpp:82]: But for non fine grained case, one operator spill usually will not block others, that is why I only handle it for fine grained executor.

File: dbms/src/Core/FineGrainedOperatorSpillContext.cpp (hunk 3 above)
Comment (SeaRise) [reply dbms/src/Core/FineGrainedOperatorSpillContext.cpp:53]: ok.

File: dbms/src/Flash/Planner/Plans/PhysicalTopN.cpp (hunk 4 above)
Comment (SeaRise) [reply dbms/src/Flash/Planner/Plans/PhysicalTopN.cpp:82]: But topN is usually together with sender, so there are similar situations.
I think it's all right.

**Review by SeaRise — APPROVED** [review]
lgtm

**Comment by SeaRise** [comment]
/hold
When you want to merge pr, comment `/hold cancel` @windtalker
